kernel/rtlil.h

```diff
@@ -695,25 +758,37 @@ struct RTLIL::Const
 	bool is_onehot(int *pos = nullptr) const;
 
 	inline RTLIL::Const extract(int offset, int len = 1, RTLIL::State padding = RTLIL::State::S0) const {
-		RTLIL::Const ret;
-		ret.bits.reserve(len);
+		bitvectorize();
+		bitvectype& bv = assert_get_bits("Const::extract");
+		bitvectype ret_bv;
+		ret_bv.reserve(len);
 		for (int i = offset; i < offset + len; i++)
-			ret.bits.push_back(i < GetSize(bits) ? bits[i] : padding);
-		return ret;
+			ret_bv.push_back(i < GetSize(bv) ? bv[i] : padding);
+		return RTLIL::Const(ret_bv);
 	}
 
 	void extu(int width) {
-		bits.resize(width, RTLIL::State::S0);
+		bitvectorize();
+		assert_get_bits("Const::extu").resize(width, RTLIL::State::S0);
 	}
 
 	void exts(int width) {
-		bits.resize(width, bits.empty() ? RTLIL::State::Sx : bits.back());
+		bitvectorize();
+		bitvectype& bv = assert_get_bits("Const::exts");
+		bv.resize(width, bv.empty() ? RTLIL::State::Sx : bv.back());
 	}
 
 	inline unsigned int hash() const {
 		unsigned int h = mkhash_init;
-		for (auto b : bits)
-			h = mkhash(h, b);
+
+		if(auto str = std::get_if<std::string>(&backing)) {
+			for (auto c : *str)
+				h = mkhash(h, c);
```

Comment:
If you have two Consts, one uses the string backing but the other uses the bit vector backing, this doesn't seem to guarantee for their hashes to be the same when their bit content is the same